### PR TITLE
feat(cli): Add ID option to 'proj role create-token' (#4632)

### LINKED ISSUE #4632: argocd proj role create-token project role - missing ability to name a token
If you are trying to resolve an environment-specific issue or have a one-off question about the edge case that does not require a feature then please consider asking a
question in argocd slack [channel](https://argoproj.github.io/community/join-slack).

Checklist:

* [x ] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [ x] I've included steps to reproduce the bug.
* [ x] I've pasted the output of `argocd version`.

**Describe the bug**

the argocd cli when creating project tokens lacks the ability to name the token (as the UI allows)

**To Reproduce**

`argocd proj role create-token project role -n`
or
`argocd proj role create-token project role token-name`

**Expected behavior**

a named token is created

**Version**

```shell
argocd: v1.7.8+ef5010c.dirty
  BuildDate: 2020-10-16T01:33:41Z
  GitCommit: ef5010c3a0b5e027fd642732d03c5b0391b1e574
  GitTreeState: dirty
  GoVersion: go1.15.2
  Compiler: gc
  Platform: darwin/amd64
argocd-server: v1.7.6+b04c25e
  BuildDate: 2020-09-19T00:52:04Z
  GitCommit: b04c25eca8f1660359e325acd4be5338719e59a0
  GitTreeState: clean
  GoVersion: go1.14.1
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: {Version:kustomize/v3.6.1 GitCommit:c97fa946d576eb6ed559f17f2ac43b3b5a8d5dbd BuildDate:2020-05-27T20:47:35Z GoOs:linux GoArch:amd64}
  Helm Version: version.BuildInfo{Version:"v3.2.0", GitCommit:"e11b7ce3b12db2941e90399e874513fbd24bcb71", GitTreeState:"clean", GoVersion:"go1.13.10"}
  Kubectl Version: v1.17.8
```

### PR DESCRIPTION
**Pull request by tetchel**
Also add some more informative output to the same command

Signed-off-by: Tim Etchells <tetchell@redhat.com>

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)). 

Fixes https://github.com/argoproj/argo-cd/issues/4632

New help:
```
Flags:
  -e, --expires-in string   Duration before the token will expire, eg "12h", "7d". (Default: No expiration)
  -h, --help                help for create-token
  -i, --id string           Token unique identifier. (Default: Random UUID)
```

New output:
```
[ /argo-cd/cmd/argocd ] 29 (master) $ ./argocd proj role create-token default tim --id tim11
Created token succeeded.
  Project: default
  Role: tim
  ID: tim11
  Expires At: Never
  Token: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJqdGkiOiJ0aW0xMSIsImlhdCI6MTYwMzM0MDM5NCwiaXNzIjoiYXJnb2NkIiwibmJmIjoxNjAzMzQwMzk0LCJzdWIiOiJwcm9qOmRlZmF1bHQ6dGltIn0.svZ7lxBfSNnZ6FHBlT-G9mjS0OJTkvPWFzdM4t0gK-o
```

If the server does not provide ID in the response (the behaviour before this PR), then just Project, Role, and the token are output; this is intended to allow compatibility with older versions of the argoCD server.

### CODE REVIEW

File: server/project/project.proto [hunk 1]
```diff
@@ -39,9 +39,10 @@ message ProjectTokenCreateRequest {
 // ProjectTokenResponse wraps the created token or returns an empty string if deleted.
 message ProjectTokenResponse {
     string token = 1;
+    int64 expiresAt = 2;
+    string id = 3;
```
Comment (jessesuen) [inline server/project/project.proto:43]: @tetchel FYI, ID and expiresAt are already encoded in the token JWT. Do you think it's too inconvenient to decode it from the token or do you feel like it needs to be surfaced in the response?

Comment (tetchel) [reply server/project/project.proto:43]: It does not look like the Expires At is encoded but I suppose I could just leave that out of this output.

image

Comment (tetchel) [reply server/project/project.proto:43]: would you like me to make that change?

Comment (jannfis) [reply server/project/project.proto:43]: The expiry time is only encoded if explicitly set - i.e. a token created with `--expires-in` option.

Comment (tetchel) [reply server/project/project.proto:43]: that make sense, a null expiry means it never expires.

OK, doing the decode on the client-side makes sense then.

**Comment by jannfis** [comment]
Just a thought about changing the output of the command - I believe that project token creation could be used in some automation scripts, and this change would break those scripts.

While I like that it prints detailed information about the created token, I'm a little unsure whether we should go ahead and change the output by default or hide it behind a `--verbose|-v` switch. Or, when we chose to make this the default, we should have an option to just print the plain token output to stdout to restore previous behaviour, and have it noted in the upgrade/release notes so people have a chance to change their scripts before upgrading.

**Comment by tetchel** [comment]
Flagging it is a good idea, you're right about breaking backwards compatibility for scripts.

I think the new output is better as the default behaviour, but I agree that it will cause pain for some users.

So either:
1. The default output is only the token and provide `--verbose` for the compete output. 
  Pro: Doesn't break existing scripts. 
  Con: The complete output will be harder to discover, and used less often.
2. The default output is the complete output and the token only is hidden behind some other flag, `--token-only|t` or `--simple|s` or something like that.
  Pro: Complete output by default
  Con: Requires a release note that scripts will break and users will have to adapt.

Let me know which option you prefer.

**Comment by jannfis** [comment]
This is a rather tough one :) Usability vs. backwards-compatibility. I know that some people actually are using means like curl or wget to the argocd-server to get its version of the CLI, and then use it with their scripts.  So they expect a fair amount of backwards compatibility, and they might not be the same people that are responsible for the life-cycle of the Argo CD installation, just plain consumers. They might have not read the release notes. 

BUT. There's a big BUT. I think this is not good practice to auto-update CLI version for scripting, and I also think this is a very small edge case with a very quick fix (i.e. adapt script to use CLI with new cmdline parameter), instead of re-writing logic (i.e. parse new output).

So I'd be fine going with option 2) and having a new switch `--token-only|t` to only dump the raw base64 encoded token to stdout. 

Other voices on this?

**Comment by tetchel** [comment]
updated:
```
[ /argo-cd/cmd/argocd ] 23 (customTokenID) $ go build . && ./argocd proj role create-token default tim --id hello --expires-in 10m
Create token succeeded for proj:default:tim.
  ID: hello
  Issued At: 2020-10-24T19:23:49-04:00
  Expires At: 2020-10-24T19:33:49-04:00
  Token: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2MDM1ODI0MjksImp0aSI6ImhlbGxvIiwiaWF0IjoxNjAzNTgxODI5LCJpc3MiOiJhcmdvY2QiLCJuYmYiOjE2MDM1ODE4MjksInN1YiI6InByb2o6ZGVmYXVsdDp0aW0ifQ.u-Pd3ecbJiQeL5qAvJb5YTQMQ11ELkw0QQzjGqd35kc
[ /argo-cd/cmd/argocd ] 24 (customTokenID) $ go build . && ./argocd proj role create-token default tim --id hello2 --expires-in 10m --token-only
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2MDM1ODI0NTMsImp0aSI6ImhlbGxvMiIsImlhdCI6MTYwMzU4MTg1MywiaXNzIjoiYXJnb2NkIiwibmJmIjoxNjAzNTgxODUzLCJzdWIiOiJwcm9qOmRlZmF1bHQ6dGltIn0.rLWYNheu4zcIyQFmoDSMIVoUlSKFoiJ8I_D7qctKFTA
```

**Review by jannfis — APPROVED** [review]
LGTM!

I marked this PR as low-impact possibly breaking change, so that we can catch up in the release notes.

**Comment by jannfis** [comment]
@jessesuen / @alexmt FYI

**Pull request merged**